lib/packets/execute.js

```diff
@@ -17,74 +17,59 @@ function isJSON(value) {
     (typeof value.toJSON === 'function' && !Buffer.isBuffer(value));
 }
 
-function toMysqlDateBuffer(value) {
-  var bufferValue = Buffer.allocUnsafe(12);
-  bufferValue.writeUInt8(11, 0);
-  bufferValue.writeUInt16LE(value.getFullYear(), 1);
-  bufferValue.writeUInt8(value.getMonth() + 1, 3);
-  bufferValue.writeUInt8(value.getDate(), 4);
-  bufferValue.writeUInt8(value.getHours(), 5);
-  bufferValue.writeUInt8(value.getMinutes(), 6);
-  bufferValue.writeUInt8(value.getSeconds(), 7);
-  bufferValue.writeUInt32LE(value.getMilliseconds() * 1000, 8);
-  return bufferValue;
-}
-
 /**
  * Converts a value to an object describing type, String/Buffer representation and length
  * @param {*} value
  */
 function toParameter(value, encoding) {
   var type = Types.VAR_STRING;
   var length;
-  var fixed = false;
+  var writer = function (value) {
+    return Packet.prototype.writeLengthCodedString.call(this, value, encoding)
+  }
   if (value !== null) {
     switch (typeof value) {
       case 'undefined':
         throw new TypeError('Bind parameters must not contain undefined');
 
       case 'number':
         type = Types.DOUBLE;
-        fixed = true;
-        var bufferValue = Buffer.allocUnsafe(8);
-        bufferValue.writeDoubleLE(value, 0);
-        value = bufferValue;
+        length = 8;
+        writer = Packet.prototype.writeDouble;
         break;
 
       case 'boolean':
+        value = value | 0;
         type = Types.TINY;
-        fixed = true;
-        var bufferValue = Buffer.allocUnsafe(1);
-        bufferValue.writeInt8(value | 0, 0);
-        value = bufferValue;
+        length = 1;
+        writer = Packet.prototype.writeInt8;
         break;
 
       case 'object':
         if (Object.prototype.toString.call(value) == '[object Date]') {
           type = Types.DATETIME;
-          fixed = true;
-          value = toMysqlDateBuffer(value);
+          length = 12;
+          writer = Packet.prototype.writeDate
         } else if (isJSON(value)) {
-          type = Types.JSON;
           value = JSON.stringify(value);
+          type = Types.JSON;
+        } else if (Buffer.isBuffer(value)) {
+          length = Packet.lengthCodedNumberLength(value.length) + value.length;
+          writer = Packet.prototype.writeLengthCodedBuffer
         }
         break;
+
```

Comment:
maybe add checks for other types to help prevent silly errors? for example, if I accidentally put callback as a parameter
I guess it's better to throw when function is passed and require explicit fn.toString() if that was actually your intention

Reply:
I've added a specific check for `function` params here: https://github.com/sidorares/node-mysql2/pull/718/files#diff-259dec4414a400a6e895f16ff1d0ca3bR724 - should preserve the stack trace better (I've added a test for it too)